cmd/snap-bootstrap/cmd_initramfs_mounts.go

```diff
@@ -220,151 +223,44 @@ func generateMountsModeRun() error {
+		if !isBaseMounted {
+			whichTypes = append(whichTypes, snap.TypeBase)
 		}
-
-		// get the primary extracted run kernel
-		kernel, err := ebl.Kernel()
-		if err != nil {
-			// we don't have a fallback kernel!
-			return fmt.Errorf("no fallback kernel snap: %v", err)
+		if !isKernelMounted {
+			whichTypes = append(whichTypes, snap.TypeKernel)
 		}
-
-		kernelFile := kernel.Filename()
-		if !validKernels[kernelFile] {
-			// we don't trust the fallback kernel!
-			return fmt.Errorf("fallback kernel snap %q is not trusted in the modeenv", kernelFile)
+		if !isSnapdMounted {
+			whichTypes = append(whichTypes, snap.TypeSnapd)
 		}
 
-		// get kernel_status
-		m, err := ebl.GetBootVars("kernel_status")
+		mounts, err := boot.InitramfsRunModeSnapsToMount(whichTypes)
 		if err != nil {
-			return fmt.Errorf("cannot get kernel_status from bootloader %s", ebl.Name())
+			return err
 		}
-
-		if m["kernel_status"] == boot.TryingStatus {
-			// check for the try kernel
-			tryKernel, err := ebl.TryKernel()
-			if err == nil {
-				tryKernelFile := tryKernel.Filename()
-				if validKernels[tryKernelFile] {
-					kernelFile = tryKernelFile
-				} else {
-					logger.Noticef("try-kernel %q is not trusted in the modeenv", tryKernelFile)
-				}
-			} else if err != bootloader.ErrNoTryKernelRef {
-				logger.Noticef("missing try-kernel, even though \"kernel_status\" is \"trying\"")
-			}
-			// if we didn't have a try kernel, but we do have kernel_status ==
-			// trying we just fallback to using the normal kernel
-			// same goes for try kernel being untrusted - we will fallback to
-			// the normal kernel snap
+		// TODO:UC20: should we have more cross-checks here?
+		if _, ok := mounts[snap.TypeBase]; ok {
+			fmt.Fprintf(stdout, "%s %s\n", mounts[snap.TypeBase], filepath.Join(dirs.RunMnt, "base"))
 		}
-
-		kernelPath := filepath.Join(dataDir, "system-data", dirs.SnapBlobDir, kernelFile)
-		fmt.Fprintf(stdout, "%s %s\n", kernelPath, filepath.Join(dirs.RunMnt, "kernel"))
-	}
-
-	// 3.1 Maybe mount the snapd snap on first boot of run-mode
-	// TODO:UC20: Make RecoverySystem empty after successful first boot
-	// somewhere in devicestate
-	if modeEnv.RecoverySystem != "" {
-		isSnapdMounted, err := osutilIsMounted(filepath.Join(dirs.RunMnt, "snapd"))
-		if err != nil {
-			return err
+		if _, ok := mounts[snap.TypeKernel]; ok {
+			fmt.Fprintf(stdout, "%s %s\n", mounts[snap.TypeKernel], filepath.Join(dirs.RunMnt, "kernel"))
 		}
-
-		if !isSnapdMounted {
-			// load the recovery system and generate mount for snapd
-			essSnaps, err := recoverySystemEssentialSnaps(seedDir, modeEnv.RecoverySystem, []snap.Type{snap.TypeSnapd})
-			if err != nil {
-				return fmt.Errorf("cannot load metadata and verify snapd snap: %v", err)
-			}
-			fmt.Fprintf(stdout, "%s %s\n", essSnaps[0].Path, filepath.Join(dirs.RunMnt, "snapd"))
+		if _, ok := mounts[snap.TypeSnapd]; ok {
+			fmt.Fprintf(stdout, "%s %s\n", mounts[snap.TypeSnapd], filepath.Join(dirs.RunMnt, "snapd"))
```

Comment:
I think for a while we can leave the snapd code here.  If we move it, it becomes asymmetric as you noticed with the install code and then it's a slightly slippery slope of changes to keep symmetry. Also is not much code or full of ifs differently from base and kernel.